src/Common/ThreadProfileEvents.cpp

```diff
@@ -518,6 +520,11 @@ void PerfEventsCounters::finalizeProfileEvents(ProfileEvents::Counters & profile
         disablePerfEvent(fd);
     }
 
+    if (disabled)
```

Comment:
for what do we need this?

Reply:
Well I think that is indeed useless if we accept that changes
-     if (auto query_context_ptr = query_context.lock())
+    if (auto global_context_ptr = global_context.lock())


Without that changes the following situation had place. Seems I mixed several fixes of one problem.

PerfEventsCounters could be
- opened and activated
- opened and disabled
- closed

When global_context->metrics_perf_events_enabled is true. Each thread opens and activates the PerfEventsCounters at attach. At detach it remains opened but becomes disabled. So the thread which is reused, do not opens already opened PerfEventsCounters, it just activate them. That is right.

When global_context->metrics_perf_events_enabled is false. Threads do not open PerfEventsCounters. just because that [is under condition](https://github.com/ClickHouse/ClickHouse/blob/73246b4788d1382e5022c6290f05676afb91b534/src/Interpreters/ThreadStatusExt.cpp#L344).
But they unconditionally call [finalizeProfileEvents](https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/ThreadStatusExt.cpp#L392).
That call is light just because `if (fd == -1)` [here](https://github.com/ClickHouse/ClickHouse/blob/73246b4788d1382e5022c6290f05676afb91b534/src/Common/ThreadProfileEvents.cpp#L516) and [here](https://github.com/ClickHouse/ClickHouse/blob/73246b4788d1382e5022c6290f05676afb91b534/src/Common/ThreadProfileEvents.cpp#L526).
That is acceptable. 

When global_context->metrics_perf_events_enabled is false but some query run with When query_context->metrics_perf_events_enabled is true.
Here I saw the problem.
Threads involved to that query calculation open and activate PerfEventsCounters. After they deactivate but leave it open. The next query will call finalizeProfileEvents, sees that events are open, read the out datated vaules.

Reply:
```
-     if (auto query_context_ptr = query_context.lock())
+    if (auto global_context_ptr = global_context.lock())
```

no objections to adding these changes ^